###Code From the *Learning the Sequence Determinants of Alternative Splicing from Millions of Random Sequences*

In an effort to make my results reproducible and aid others in their own data analysis, I've included all of the code used in the paper. I hope this will be useful and feel free to conact me with suggestions/comments/questions. Happy data mining!

-Alex

#### ipython notebooks:
[Notebook 0A: Fastq to Isoform Counts (Alt. 3SS)](ipython.notebooks/Cell2015_N0A_A3SS_Fastq_to_Spliced_Reads.ipynb)<br>
[Notebook 0B: Fastq to Isoform Counts (Alt. 5SS)](ipython.notebooks/Cell2015_N0B_A5SS_Fastq_to_Spliced_Reads.ipynb)<br>
[Notebook 1: Library Statistics](ipython.notebooks/Cell2015_N1_Library_Statistics.ipynb)<br>
[Notebook 2: Splice Site Analysis](ipython.notebooks/Cell2015_N2_Splice_Site_Analysis.ipynb)<br>
[Notebook 3: Splicing Frame and Nonsense Mediated Decay Analysis](ipython.notebooks/Cell2015_N3_A5SS_Splicing_Frame_Analysis.ipynb)<br>
[Notebook 4: Estimating Motif Effects](ipython.notebooks/Cell2015_N4_Motif_Effect_Sizes.ipynb)<br>
[Notebook 5: Combinatorial Motif Effects](ipython.notebooks/Cell2015_N5_Combinatorial_Motif_Effects.ipynb)<br>
[Notebook 6: Learning Curves for Models Predicting Alt. 5SS Usage](ipython.notebooks/Cell2015_N6_A5SS_Model_Learning_Curves.ipynb)<br>
[Notebook 7: A Model of Alternative 5\prime Splicing](ipython.notebooks/Cell2015_N7_A5SS_Model.ipynb)<br>
[Notebook 8: Predictions on Alt. 5SS Events in the Genome](ipython.notebooks/Cell2015_N8_HAL_Genome_Predictions.ipynb)<br>
[Notebook 9: Training a Joint Model with Both the Alt. 5SS and Alt. 3SS Libraries](ipython.notebooks/Cell2015_N9_Training_Joint_A5SS_A3SS_Model.ipynb)<br>
[Notebook 10: Predicting the Effects of SNPs on Alternative 5\prime Splicing](ipython.notebooks/Cell2015_N10_A5SS_SNP_Prediction.ipynb)<br>
[Notebook 11: Predicting the Effects of SNPs on Exon Skipping](ipython.notebooks/Cell2015_N11_Predicting_Cassette_Exon_SNP_Effects.ipynb)<br>







